sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/AnsiTypeCoercion.scala

```diff
@@ -0,0 +1,316 @@
+      StringLiteralCoercion ::
+      Nil
+
+  /**
+   * Find the tightest common type of two types that might be used in a binary expression.
+   */
+  override def findTightestCommonType(t1: DataType, t2: DataType): Option[DataType] = {
+    (t1, t2) match {
+      case (t1, t2) if t1 == t2 => Some(t1)
+      case (NullType, t1) => Some(t1)
+      case (t1, NullType) => Some(t1)
+
+      case (t1: NumericType, t2: NumericType) =>
+        findTightestCommonNumericType(t1, t2)
+
+      case (_: TimestampType, _: DateType) | (_: DateType, _: TimestampType) =>
+        Some(TimestampType)
+
+      case (t1, t2) => findTypeForComplex(t1, t2, findTightestCommonType)
+    }
+  }
+
+  @tailrec
+  private def findTightestCommonNumericType(t1: NumericType, t2: NumericType): Option[DataType] = {
+    (t1, t2) match {
+      case (i: IntegralType, d: DecimalType) =>
+        if (d.isWiderThan(i)) {
+          Some(t2)
+        } else {
+          findTightestCommonNumericType(DecimalType.forType(i), d)
+        }
+
+      case (t1: DecimalType, t2: IntegralType) =>
+        findTightestCommonNumericType(t2, t1)
+
+      case (t1: DecimalType, t2: DecimalType) =>
+        Some(DecimalPrecision.widerDecimalType(t1, t2))
+
+      case (_: FractionalType, _: DecimalType) | (_: DecimalType, _: FractionalType) =>
+        Some(DoubleType)
+
+      // Promote numeric types to the highest of the two
+      case _ =>
+        // The cases that t1 or t2 is DecimalType should be handled already.
+        assert(!t1.isInstanceOf[DecimalType] && !t2.isInstanceOf[DecimalType])
+        val index = numericPrecedence.lastIndexWhere(t => t == t1 || t == t2)
+        val widerType = numericPrecedence(index)
+        if (widerType == FloatType) {
+          // If the input type is an Integral type and a Float type, simply return Double type as
+          // the tightest common type to avoid potential precision loss on converting the Integral
+          // type as Float type.
+          Some(DoubleType)
+        } else {
+          Some(widerType)
+        }
+    }
+  }
+
+  override def findWiderTypeForTwo(t1: DataType, t2: DataType): Option[DataType] = {
+    findTightestCommonType(t1, t2)
+  }
+
+  override def findWiderCommonType(types: Seq[DataType]): Option[DataType] = {
+    types.foldLeft[Option[DataType]](Some(NullType))((r, c) =>
+      r match {
+        case Some(d) => findWiderTypeForTwo(d, c)
+        case _ => None
+      })
+  }
+
+  override def implicitCast(e: Expression, expectedType: AbstractDataType): Option[Expression] = {
+    (e, expectedType) match {
+      // This type coercion system will allow implicit converting String type literals as other
+      // primitive types, in case of breaking too many existing Spark SQL queries.
+      case (_ @ StringType(), a: AtomicType) if e.foldable && a != BooleanType && a != StringType =>
+        Some(Cast(e, a))
+
+      case (_ @ StringType(), NumericType) if e.foldable =>
+        Some(Cast(e, DoubleType))
```

Comment:
We need this entry? What's a difference between this entry and the first entry (L171)?

Reply:
Yes, we need. This one can match the object `NumericType`, which means any numeric type.

Reply:
All the numeric types are also atomic types, so this match cannot happen? Am I missing something?